### DISCUSSION TITLE
Avoid Set-PnPClientSideWebPart creating a new version

### DISCUSSION THREAD

**Discussion by p1995** [pnp/powershell#158] [Q&A]
I'm performing a content migration from classic SharePoint page (SharePoint Online Classic Site. Let's call it source) to Modern site pages(Let's call it destination). In the Source there are many versions so I store the result in the form of  array with each item being the information of each version. So In PnP-Powershell i  make  use of a for loop to push data each iteration. Each iteration, I Update the properties of the Custom content Editor Webpart using **Set-PnPClientSideWebPart** cmdlet and update the site page columns (like modified date, modified by etc.) using **Set-PnPListItem** cmdlet. The real issue here is these two cmdlets are creating 2 versions in site Page i.e. if there are 5 versions in source there will be 10 versions in destination. I'm aware I can use -SystemUpdate option in **Set-PnPListItem** but the Modified by and date columns remains unchanged. So my ultimate requirement is in each iteration I want a new version with updated Webpart properties and modified fields set to the values passed through the PowerShell. Can I  pass -SystemUpdate in **Set-PnPClientSideWebPart** cmdlet or something similar that prevents a new version creation?

These two lines of snippet I have been using inside a for loop `below`
```
Set-PnPClientSideWebPart -Page $pagename -Identity 'CustomContentEditor' -PropertiesJson $propertyObjectInJson  (This is creating a new version)

$updateSitePageFields=Set-PnPListItem -List "SitePages"  -Identity $page.PageId -SystemUpdate -Values @{"Modified"=$array[$i].Modified;"Author"=$array[$i].Author.Email;"Created"=$array[$i].Created;"Title"=$array[$i].Title;"DocumentType1"=$array[$i].DocumentType1;"Editor"=$array[$i].Editor.Email} (Currently I have set  -SystemUpdate so a version is not being created on execution)
```
Please help me with this issue

P.S.:  If there a solution to [#2916](https://github.com/pnp/PnP-PowerShell/issues/2016) and [#2016](https://github.com/pnp/PnP-PowerShell/issues/2016), then this will also be the answer to my question

**Comment by ToddKlindt** [comment]
Are you using the [Page Modernization](https://docs.microsoft.com/en-us/sharepoint/dev/transform/modernize-userinterface-site-pages-powershell) PowerShell to create the new pages? If so, you can do a lot with the mapping files. You might be able to get everything you need done with them removing the need for the Set-PnPListItem step that is creating another version.

**Reply by p1995** [reply]
In my case I should not modify the Source Site. The solution you provided seems it converts the pages to modern site pages in the source. with those 2 cmdlets creating 2 versions, I'm deleting the version created by **Set-PnPClientSidewebpart** in the end but again this results in alternate numbering in the versions. Any suggestions?

P.S. Any Workaround for UpdateOverwriteVersion in pnp Powershell

**Reply by jansenbe** [reply]
You can also do cross site transformation, read the source page (nothing will change) and create a modern version from it your target site. The source site can be in SharePoint Online or SharePoint on-premises while the target site has to be in SharePoint Online.

See https://docs.microsoft.com/en-us/sharepoint/dev/transform/modernize-userinterface-site-pages-powershell. Use the -TargetWebUrl or -TargetConnection properties on ConvertTo-PnPPage (note that docs still mentioned old PS cmdlets, we're updating things currently).

**Reply by p1995** [reply]
Yes @jansenbe, I can use modernization of pages but I have challenges here:
I have a content editor (which is customized using page layout) in source page but in destination after modernization, I should be able to show the data of each version in the custom created Rich Text Editor not on Out of the box webpart like Text etc. since it is lacking options like insert image and Table column merging, splitting etc.
 
It wouldn't have been a challenge if there was an option in **Set-PnPListItem** to overwrite the existing version but **not** mess the Editor fields which -SystemUpdate fails to do

**Reply by erwinvanhunen** [reply]
Updating author and editor fields is a bit weird in CSOM. In order to set either field you -also- have to set the other. And then, systemupdate will allow you to update the author field, but not the editor field, for that you need UpdateOverwriteVersion. 

We have updated the Set-PnPListItem cmdlet now to handle the field management behind the scenes. E.g. if you set the editor field, we will also retrieve the values for the existing item for the author field (and created and modified) and reinclude those values in the request. This means that if you change the editor field, actually 3 additional fields will be set to the value they already had. 

We also updated the cmdlet to support -UpdateType, instead of -SystemUpdate (the latter is marked as obsolete now, but it will still work). So if you want to set author and / or editor:

```powershell
Set-PnPListItem -List [yourlist] -Id [id] -Values @{"Author"="yourauthor@domain.com"} -UpdateOverwriteVersion
``` 
or
```powershell
Set-PnPListItem -List [yourlist] -Id [id] -Values @{"Author"="yourauthor@domain.com";"Editor"="youreditor@domain.com"} -UpdateOverwriteVersion
```
etc. etc.

This change will be available in the next nightly build: 1.2.6-nightly.

**Reply by p1995** [reply]
Thank You @erwinvanhunen. This feature was much needed as it has been remained challenging in the past. Until this feature rolls out, I found a way to combat this issue. This is a part of the snippet written in  for loop

```
$updateWebpartProperty=Set-PnPClientSideWebPart -Page $pagename -Identity 'custom-Rich-Text -PropertiesJson $propertyObjectInJson
Set-PnPList -Identity "SitePages"  -EnableMinorVersions $false //Default this command creates a minor version. Setting this will result **Set-PnPListItem** creating a major version
$updateSitePageFields=Set-PnPListItem -List "SitePages"  -Identity $page.PageId -Values @{"Modified"=$array[$i].Modified;"Editor"=$array[$i].Editor.Email;"Author"=$array[$i].Author.Email;"Created"=$array[$i].Created;"Title"=$array[$i].Title;"DocumentType1"=$array[$i].DocumentType1}
Set-PnPList -Identity "SitePages"  -EnableMinorVersions $true // so that **Set-PnPClientSideWebpart** creates a minor version for next iteration
Remove-PnPFileVersion -Url $FileRelativeURL  -Identity $file[$file.count-2].Id -Force // Deleting the minor version created by **Set-PnPClientSideWebpart**
```

Meanwhile, Hope this script help!!
@ToddKlindt @jansenbe @erwinvanhunen  Let me know your feedback on this script.

**Reply by erwinvanhunen** [reply]
Interesting approach on the script! If it works for it, it works :-) Btw, the 1.2.6-nightly release is available for you try out:

```powershell
Update-Module -Name PnP.PowerShell -AllowPrerelease
```

**Reply by erwinvanhunen** [reply]
If it works, please mark this post as answered, it will help us to keep track of things.

**Reply by p1995** [reply]
Thanks for the follow-up. I executed Update-module command as you suggested and verified. I found out **-UpdateOverwriteVersion** still not working and it is not appearing in the intellisesne. I even tried adding that parameter to **Set-PnPListItem** but it throws an error (`Set-PnPListItem : A parameter cannot be found that matches parameter name 'UpdateOverwriteVersion'.`
).

**Reply by erwinvanhunen** [reply]
Sorry, I wrote the message before actually released it and we implemented it differently:

Set-PnPListItem -List [list] -Id [id] -Values @[values] -UpdateType UpdateOverwriteVersion

**Reply by erwinvanhunen** [reply]
The UpdateType parameter takes 'Update' (the default), 'SystemUpdate' or 'UpdateOverwriteVersion'

**Reply by p1995** [reply]
**UpdateOverwriteVersion** is appearing in intellisense but it does nothing. It is as same as not passing any parameters as it created minor versions each time.

**Reply by p1995** [reply]
> Sorry, I wrote the message before actually released it and we implemented it differently:
> 
> Set-PnPListItem -List [list] -Id [id] -Values @[values] -UpdateType UpdateOverwriteVersion

Also there is no Parameter called Id. As it is not appearing in intellisense. Do you mean **-Identity**?

**Reply by erwinvanhunen** [reply]
Yes. I did it by heart when I wrote the example. There are over 500 cmdlets, it's sometimes hard to remember all the details correct.
